compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -213,6 +213,65 @@ impl<'a> AstValidator<'a> {
         err.emit();
     }
 
+    fn check_anonymous_field(&self, field: &FieldDef) {
+        let FieldDef { ty, ident, .. } = field;
+        match &ty.kind {
+            TyKind::AnonymousStruct(..) | TyKind::AnonymousUnion(..) => {
+                if let Some(Ident { name: kw::Underscore, .. }) = ident {
+                    // Empty because we want to error only if Ident != `_`
+                } else {
+                    let struct_kind = if matches!(ty.kind, TyKind::AnonymousStruct(..)) {
+                        "struct"
+                    } else {
+                        "union"
+                    };
+                    let msg =
+                        format!("anonymous {} not allowed with a named identifier", struct_kind);
+                    self.err_handler()
+                        .struct_span_err(field.span, msg.as_str())
+                        // There are no `None` idents on a struct field def, so it should be safe to unwrap
+                        .span_suggestion(
+                            ident.unwrap().span,
+                            "replace with",
+                            "_".to_string(),
+                            Applicability::MachineApplicable,
+                        )
+                        .emit();
+                }
+            }
+            TyKind::Path(..) => {
+                // If the anonymous field contains a Path as type, we can't determine if the path is a valid
+                // struct or union, so we skip the check
+            }
+            _ => {
+                if let Some(Ident { name: kw::Underscore, .. }) = ident {
+                    let msg = "only anonymous structs and unions are allowed";
+                    let label = "not an anonymous struct or union";
+                    self.err_handler()
+                        .struct_span_err(field.span, msg)
+                        .span_label(ty.span, label)
+                        .emit();
+                }
+            }
+        }
+    }
+
+    fn deny_anonymous_struct(&self, ty: &Ty, parent_span: Span, where_: &str) {
```

Comment:
Calls to `deny_anonymous_struct` are missing in numerous locations now, for example `struct { ... }` types won't be rejected in any nested type contexts like `[struct { field: u8 }; 1]`.

It's not realistic to catch all the cases manually.
Instead there should be a single call to `deny_anonymous_struct` in `fn visit_ty`, but then `fn visit_ty` should be skipped (in favor of calling `AstValidator` logic + `walk_ty` manually) in places where anonymous types are *allowed*, i.e. anonymous fields in structs and unions.

Reply:
Yeah, that's a better implementation approach I think. The error messages will be less descriptive though, because I don't think I can get the parent `Span` of a type from `fn visit_ty`.